Fix buggy move messages count problem.

whilstsomebody opened the pull request:
This PR fixes the bug introduced by #32155 .

The count of messages to be moved was observed to be incorrect occasionally because the function named `messages_util.get_messages_in_topic` was used which relied on the `recent conversations` cache for the history.
    
Now, we introduced a new function named `get_messages_in_view_currenlty_focused` to fetch the count specifically for the view in focus. If we don't have this, then we use `get_messages_in_topic` as before.

---

For the different cases when the user opens the **Move Messages** modal from the left sidebar, we implement `get_count_of_messages_in_view_currently_focused`, which counts the messages in two scenarios:  

- **When the currently focused view is "stream," and the user opens the message modal for the topic they are currently focused on.**  
  In this case, we retrieve the messages of that topic using `get_messages_in_view_currently_focused`.  

- **When the currently focused view is "stream," and the user opens the message modal for a different topic within the same stream.**  
  In this case, `get_messages_in_view_currently_focused` fails to give the message count correctly. This is exactly the bug mentioned by @gnprice in [#issues > 📂 wrong message count in move-messages dialog](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.93.82.20wrong.20message.20count.20in.20move-messages.20dialog/with/2208651).  
To handle this case, we use the `get_messages_in_topic` function instead to obtain the correct message count.  

- **Difference between the two functions:**  
  `get_messages_in_view_currently_focused` retrieves messages for the topic that is currently in focus, whereas `get_messages_in_topic` fetches messages for the selected topic, regardless of the current focus.  

---
**Note: I have also changed the count display text from `Atleast N messages will be moved.` to `N+ messages will be moved` in a separate commit.** 

For more info, see [CZO](https://chat.zulip.org/#narrow/channel/9-issues/topic/wrong.20message.20count.20in.20move-messages.20dialog/near/2025435).

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@sahil839 could you please take a look at this one (if @timabbott doesn't get to it first)?

whilstsomebody commented on web/src/stream_popover.ts:
```diff
@@ -340,6 +340,103 @@ async function get_message_placement_in_conversation(
+// both message_lists.current.data.fetch_status.has_found_newest
+// and message_lists.current.data.fetch_status.has_found_oldest are true;
+// otherwise, we cannot be certain of the correct count.
+export function get_count_of_messages_to_be_moved(
```
These functions have been extracted out of `build_move_topic_to_stream_popove` function for testing purpose.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -340,6 +340,105 @@ async function get_message_placement_in_conversation(
+    if (selected_option === "change_one") {
+        return 1;
+    }
+    if (selected_option === "change_later" && message_id !== undefined) {
```
I think `assert(message_id !== undefined)` would be better here instead having `&&` block in if condition, since we allow selecting the "change_later" option only if moving the message from message actions popover.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -340,6 +340,105 @@ async function get_message_placement_in_conversation(
+        topic_name,
+    );
+    if (messages_for_currently_focused_view.length > 0) {
+        return messages_for_currently_focused_view.length;
```
The current logic should be correct, but maybe we can write the code in a way that checks whether the message's current stream and topic matches the narrowed view. Using that, we can use `message_lists.current?.data.all_messages()` if the above match occurs  (with additional filtering needed if the narrowed view is a stream narrow) and use `message_util.get_messages_in_topic` otherwise.

I guess that would make the code much easier to read and understand.

whilstsomebody replied:
There's now a single `current_view_contains_topic()` helper, narrowed to the topic, or to the
channel containing it — and the count picks its source from it:

```ts
const loaded_messages_in_topic = current_view_contains_topic(current_stream_id, topic_name)
    ? message_util.get_loaded_messages_in_topic_in_current_view(current_stream_id, topic_name)
    : message_util.get_loaded_messages_in_topic(current_stream_id, topic_name);
```

`update_move_messages_count_text` uses the same helper for the exact vs.
"N+" wording, in place of the `can_have_exact_count_in_narrow` block it
computed itself, so the count and the wording can't disagree.

whilstsomebody committed:
test: Test current get_count_of_messages_to_be_moved() function.

whilstsomebody committed:
move_topic_popover: Fix the buggy message count problem.

The count of messages to be moved was observed to be
incorrect occasionally because the function named
"messages_util.get_messages_in_topic" was used which
relied on the "recent conversations" cache for the
history.

Now, we try to fetch the messages for the current view
and update the count accordingly. In case we don't have
those messages, then we directly fetch the messages for
the topic using "get_messages_in_topic".

Fixes the bug introduced by #32155.

whilstsomebody commented on web/src/stream_popover.ts:
```diff
@@ -340,6 +340,87 @@ async function get_message_placement_in_conversation(
+// both message_lists.current.data.fetch_status.has_found_newest
+// and message_lists.current.data.fetch_status.has_found_oldest are true;
+// otherwise, we cannot be certain of the correct count.
+export function get_count_of_messages_to_be_moved(
```
This function has been extracted of the `build_move_topic_to_stream_popover(` so that it can be tested easily.

whilstsomebody commented on web/src/stream_popover.ts:
```diff
@@ -340,6 +340,87 @@ async function get_message_placement_in_conversation(
+    return message_util.get_count_of_messages_in_topic(current_stream_id, topic_name);
+}
+
+export function update_move_messages_count_text(
```
This function has been extracted of the `build_move_topic_to_stream_popover` so that it can be tested easily.

whilstsomebody commented:
Hey @timabbott , could you please take a look ?

timabbott commented on web/src/message_util.ts:
```diff
@@ -24,17 +24,52 @@ export function do_unread_count_updates(messages: Message[], expect_no_new_unrea
+    stream_id: number,
+    topic: string,
+): Message[] {
+    const current_view_messages = message_lists.current?.data.all_messages();
```
I don't understand the algorithm here -- can you explain it?

whilstsomebody replied:
Explained in the PR description above.

timabbott commented:
You'll want to add a commit reverting 9107101fe0d17cb7872475c17cea8eee8d55152a to this PR.

whilstsomebody commented:
@zulipbot remove "has conflicts"

timabbott commented on web/tests/message_edit.test.cjs:
```diff
@@ -4,15 +4,26 @@ const assert = require("node:assert/strict");
+    data: {
+        fetch_status: {},
+    },
+};
```
In general this test uses a lot more mocking than would be ideal. Better would be to just popular objects in `stream_data` using the APIs in `web/tests/lib/example_stream.ts`, and similarly for other objects.

But it's still usable.

timabbott replied:
Actually, I take that back -- I'm not convinced your tests here are actually testing any of the logic that we care about here for correctness -- it seems like you're mocking the code that would be most likely to be wrong in each case. 

I recommend (1) adding your new tests in a new test file, `message_move_counts.test.cjs`, and (2) writing that test without using `mock_esm` -- just create objects using the relevant APIs, so that we're actually exercising the real code.

timabbott commented on web/src/message_util.ts:
```diff
@@ -24,13 +24,52 @@ export function do_unread_count_updates(messages: Message[], expect_no_new_unrea
 ): number {
-    const all_messages = get_loaded_messages_in_topic(stream_id, topic);
-    return all_messages.filter((msg) => msg.id >= message_id).length;
+    let all_messages = get_messages_in_view_currently_focused(stream_id, topic);
```
Don't call this `all_messages`! It's not "all" of anything natural. I'd call this

`const loaded_messages_in_current_view = ...`

and then the other block would be `const loaded_messages_in_all_messages` or something -- they're totally unrelated and the `let` here reusing the same variable name is also a bad pattern.

whilstsomebody committed:
popovers: Enable move-messages count.

We had hidden the move-messages count in
9107101fe0d17cb7872475c17cea8eee8d55152a due to
incorrect count caused by logic error. Since the
logic has now been corrected, we are reverting it
in this commit.

whilstsomebody commented on web/tests/message_move_count.test.cjs:
```diff
@@ -0,0 +1,231 @@
+    );
+    assert.equal(count, 1);
+
+    count = stream_popover.get_count_of_messages_to_be_moved(
```
Should I also assert the functions working behind this function to get the count in both the commits? I believe that would be more helpful to show the actual difference between the logics in both commits.

whilstsomebody commented:
The most important difference between the previous and the current logic is:

- In the previous logic, we were calculating the count with the help of `all_messages_data.all_messages()` in each narrow.
- In the current logic, we are calculating the count with the help of `current.data.all_messages()` which gives the loaded messages in the current narrow.

whilstsomebody commented:
@zulipbot remove "has conflicts"

whilstsomebody commented:
Hey @timabbott , could you review this PR again. Thanks!

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -375,6 +375,87 @@ async function get_message_placement_in_conversation(
+    } else {
+        message_text = $t(
+            {
+                defaultMessage: "{count}+ messages will be moved.",
```
It is not clear what this commit is for. The commit message says it adds test, so the changes to move this function to be at module level makes sense but this text is also changes the string here.

whilstsomebody replied:
Split out the string change is now its own commit after the bug fix, so this one only extracts the two functions to module level so they can be tested. Its wording is unchanged from what's on main; the only thing that happens to that line here is the reindent from the move.

sahil839 commented on web/tests/stream_popover.test.cjs:
I don't think we currently have a test file for testing a single thing. We usually have test files corresponding to the `src` file, so we should maintain that.

whilstsomebody replied:
Renamed it to `stream_popover.test.cjs`, both functions under test (`get_count_of_messages_to_be_moved` and `update_move_messages_count_text`) are in `stream_popover.ts`, and there was no test file for that module yet. I folded the rename into the commit that adds the file, so it isn't added
under one name and renamed a commit later.

@timabbott had asked in https://github.com/zulip/zulip/pull/32892#discussion_r2170407881 for these tests to move out of `message_edit.test.cjs` into a file of their own, so naming that file after the src module covers both.

sahil839 commented on web/src/message_util.ts:
```diff
@@ -72,17 +72,60 @@ export function do_unread_count_updates(messages: Message[], expect_no_new_unrea
 
-export function get_count_of_messages_in_topic(stream_id: number, topic: string): number {
-    return get_loaded_messages_in_topic(stream_id, topic).length;
+export function get_loaded_messages_in_view_currently_focused(
```
This name is confusing as we also filter by stream and topic later and not just return the messages in view.

whilstsomebody replied:
`get_loaded_messages_in_topic_in_current_view`

sahil839 commented on web/src/message_util.ts:
```diff
@@ -72,17 +72,60 @@ export function do_unread_count_updates(messages: Message[], expect_no_new_unrea
 ): number {
-    const all_messages = get_loaded_messages_in_topic(stream_id, topic);
-    return all_messages.filter((msg) => msg.id >= message_id).length;
+    const loaded_messages_in_current_view = get_loaded_messages_in_view_currently_focused(
```
I don't get how this helps. Maybe I don't have the complete context, but won't this only show count of messages in the view if say we are moving messages from a view like search view while more messages will be moved. Or we are fine as we would show count like "2+" or "At least 2" whichever has been decided.

sahil839 commented on web/styles/popovers.css:
```diff
@@ -748,8 +748,15 @@ ul.popover-group-menu-member-list {
+        /* For cases where the count isn't displayed, which may primarily
+           occur due to bugs, avoid applying the above negative margin. */
+        &:empty {
+            margin-top: unset;
```
I think this is not the expected look of the count text.

<img width="1414" height="976" alt="Image" src="https://github.com/user-attachments/assets/8644b048-213d-4838-b8fa-7e372cc5bc2c" />

whilstsomebody replied:
<img width="603" height="466" alt="image" src="https://github.com/user-attachments/assets/10080a6f-5f26-4475-8cd1-c31b1c857330" />

Fixed the italics, not sure about the `top-margin`(followed from the mark message as read model).

sahil839 replied:
I only had concerns about space between the count text and checkbox below it.

whilstsomebody replied:
Should we follow the mar message read model for this?

sahil839 replied:
I am not sure. Maybe you can post in #design for feedback.

sahil839 reviewed (commented):
Posted some comments.

whilstsomebody commented:
@sahil839 , PTAL. Some of the above comments have been addressed using claude.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -376,10 +377,28 @@ async function get_message_placement_in_conversation(
+//
+// Containing the topic does not mean the view has fetched all of its
+// messages; `fetch_status` answers that question.
+function current_view_contains_topic(current_stream_id: number, topic_name: string): boolean {
```
Should this function also mention something about stream also being required to match the stream ID?

whilstsomebody replied:
Yes , the comment now leads with it, and matching the channel is the function's first checK;

```ts
// The current message list contains the topic being moved when it is
// narrowed to that topic, or to the channel the topic is in. ...
function current_message_list_contains_topic(stream_id: number, topic_name: string): boolean {
    const current_filter = message_lists.current?.data.filter;
    if (current_filter === undefined || narrow_state.stream_id(current_filter) !== stream_id) {
        return false;
    }
```

The function is also renamed from `current_view_contains_topic`, since what it really asks about is the current message list.

(see my reply about the "0+" count below.)

sahil839 commented on web/src/message_util.ts:
```diff
@@ -81,6 +81,10 @@ export function get_count_of_messages_in_topic_sent_after_current_message(
     return all_messages.filter((msg) => msg.id >= message_id).length;
 }
 
+export function get_count_of_messages_in_topic(stream_id: number, topic: string): number {
```
This is being removed in second commit, so I think we can just avoid adding it here and keep using `get_loaded_messages_in_topic(stream_id, topic).length` directly.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -376,6 +376,88 @@ async function get_message_placement_in_conversation(
+}
+
+export function update_move_messages_count_text(
+    selected_option: string,
```
Now that `selected_option` is used in module-level functions, I think we can use more stricter type for this that contains only three suported values.

whilstsomebody replied:
Added `PropagateMode` in the commit that moves these functions to module level, so the parameter never has the loose type:

```ts
const propagate_mode_schema = z.enum(["change_one", "change_later", "change_all"]);
type PropagateMode = z.infer<typeof propagate_mode_schema>;
```

It also made `update_move_messages_count_text` simpler: with the union, `change_one` and `change_later` handled, the remaining else is `change_all` without a second string comparison.

sahil839 commented on web/tests/stream_popover.test.cjs:
```diff
@@ -0,0 +1,244 @@
+    });
+}
+
+run_test("test_get_count_of_messages_to_be_moved", ({override}) => {
```
`test_` prefixes can be removed here.

sahil839 commented on web/tests/stream_popover.test.cjs:
```diff
@@ -0,0 +1,244 @@
+const people = zrequire("people");
+const stream_data = zrequire("stream_data");
+
+const all_messages_data = mock_esm("../src/all_messages_data");
```
This test file does not pass on first commit.

whilstsomebody replied:
Rewritten so the first commit passes while demonstrating the bug.

sahil839 commented on web/tests/stream_popover.test.cjs:
```diff
@@ -0,0 +1,244 @@
+
+const alice = make_user({email: "alice@example.com", user_id: 1, full_name: "Alice"});
+const bob = make_user({email: "bob@example.com", user_id: 2, full_name: "Bob"});
+const carol = make_user({email: "carol@example.com", user_id: 3, full_name: "Carol"});
```
Are these users required in the tests below?

whilstsomebody replied:
Only one is required. 

```js
// Adding messages to a message list records their senders as
// participants of the conversation, so the sender has to be a user we
// know about.
const sender = make_user();
people.add_active_user(sender);
```

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -405,28 +429,21 @@ export function update_move_messages_count_text(
-            }
+    } else if (current_view_contains_topic(current_stream_id, topic_name)) {
+        const has_found_newest = message_lists.current?.data.fetch_status.has_found_newest();
+        const has_found_oldest = message_lists.current?.data.fetch_status.has_found_oldest();
```
How do we handle cases where user does not have access to the full history of the channel?

whilstsomebody replied:
Hmm, 

We were claiming an exact count there. `has_found_oldest()` is true once the server stops sending older messages, whether that's because there are none or because the channel's history is protected; `fetch_status.history_limited()` distinguishes them. A move takes the messages we can't see, so "move all" now requires it to be false:

```ts
exact_message_count =
    fetch_status.has_found_newest() &&
    fetch_status.has_found_oldest() &&
    !fetch_status.history_limited();
```

"This and all following messages" stays exact, since the truncation is only on the older side and `has_found_newest()` already covers the newer one.

sahil839 replied:
But I think this will show users that there are additional messages in the topic which they cannot access, I don't think we want to do that. So, am not sure what is the correct behavior here.

whilstsomebody replied:
Ohh sorry, the above is wrong. 
I checked it again, and found that `history_limited` isn't about channel history access, the API docs define it as "limited due to plan restrictions", from the realm's `first_visible_message_id`. Those messages move either way, so I think "+" is fine there.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -376,10 +377,28 @@ async function get_message_placement_in_conversation(
+        return false;
+    }
+
+    if (narrow_state.narrowed_by_stream_reply()) {
```
Currently when trying to move topic from a channel feed view (http://localhost:9991/#topics/channel/13-social) shows count as "0+".

whilstsomebody replied:
Okh so this was my mistake. 

The problem was that the two halves of the logic were looking at different things. 

The check used `narrow_state`, but the count came from `message_lists.current`. In the channel topics view those disagree as there's no message list at all (`views_util.show()` sets `message_lists.current` to undefined), but `narrow_state.filter()` still returns the inbox's channel filter. So the check thought we were in a channel narrow and could count exactly, while the count itself had nothing to look at so 0, and "0+" since `fetch_status` was undefined too.

Now both come from `message_lists.current.data.filter`, so a view without a message list can't claim to contain the topic. Those views fall back to the recent conversations cache, like before my change. There's a test for it now covering the inbox and an interleaved view.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -376,10 +377,28 @@ async function get_message_placement_in_conversation(
+        return !user_topics.is_topic_muted(current_stream_id, topic_name);
+    }
+
+    return narrow_state.narrowed_by_topic_reply() && narrow_state.topic() === topic_name;
```
For topic narrows with near links, the count is shown as "count+" even when exact count can be shown.

whilstsomebody replied:
Fixed!

`narrow_state.narrowed_by_topic_reply()` filters out only with terms, so a `channel/topic/near` narrow has three terms and fails its `terms.length === 2` check. `Filter.is_conversation_view()` already treats `["channel", "topic", "near"]` and `["channel", "topic", "with"]` as conversation views, so the check uses that instead:

```ts
return (
    current_filter.is_conversation_view() &&
    util.lower_same(narrow_state.topic(current_filter), topic_name)
);

```

sahil839 reviewed (commented):
Posted some comments.

whilstsomebody commented:
@sahil839 , I have addressed the above comments. PTAL.

sahil839 commented on web/tests/stream_popover.test.cjs:
```diff
@@ -0,0 +1,203 @@
+
+    // Wrong: two of the topic's messages are at or after the selected
+    // one, and this view has them loaded, but the count comes from the
+    // recent conversations cache, which holds only the newest one.
```
I am not sure of the way these tests are returned. We add data to the recent conversations cache ourselves, which can contain all the topic messages as well practically and then mention that the count being asserted is wrong.

whilstsomebody replied:
I have reworked it

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -380,10 +380,33 @@ async function get_message_placement_in_conversation(
+//
+// Containing the topic is not the same as having fetched all of it;
+// `fetch_status` answers that question.
+function current_message_list_contains_topic(stream_id: number, topic_name: string): boolean {
```
I think this will return false for a channel narrow with `near` part.

sahil839 commented on web/src/stream_popover.ts:
```diff
@@ -393,17 +416,26 @@ export function get_count_of_messages_to_be_moved(
+    // history it has fetched for the topic, and fall back to the recent
+    // conversations cache, which is missing the older messages of a
+    // topic that has not been active recently.
+    const loaded_messages_in_topic = current_message_list_contains_topic(
```
This function is called twice. Can we deduplicate the code?

whilstsomebody replied:
Done. 
It was only called once, but it and `update_move_messages_count_text` were both calling `current_message_list_contains_topic` with the same arguments to decide the same thing. They're now one function, `get_move_messages_count`, returning `{count, is_exact}`
